pkg/common/diskutil/file.go

```diff
@@ -3,16 +3,39 @@ package diskutil
 import (
 	"io/ioutil"
 	"os"
+	"path/filepath"
 )
 
+// AtomicWriteFile writes data out.  It writes to a temp file first, fsyncs that file,
+// then swaps the file in.  os.Rename is an atomic operation, so this sequence avoids having
+// a partially written file at the final location.  Finally, fsync is called on the directory
+// to ensure the rename is persisted.
 func AtomicWriteFile(path string, data []byte, mode os.FileMode) error {
-	if err := ioutil.WriteFile(path+".tmp", data, mode); err != nil {
+	tmpPath := path + ".tmp"
+	if err := ioutil.WriteFile(tmpPath, data, mode); err != nil {
 		return err
 	}
 
-	if err := os.Rename(path+".tmp", path); err != nil {
+	if err := sync(tmpPath); err != nil {
```

Comment:
I did some reading on this out of curiosity - turns out that calling `fsync()` on a file that has already been closed doesn't provide the same guarantees as calling `fsync()` on a file after `write()` but before `close()`. This makes some sense, because `fsync` operates on a file descriptor, not a file itself.

This [SO](https://stackoverflow.com/a/50158433) post has some good info plus appropriate LMKL links. Also seems somewhat related to that postgres fsync issue a year or so ago.

Reply:
That's a really good catch. I'll fix to use the same fd.